### PR TITLE
Allocate guest user IDs numericcally

### PR DESCRIPTION
**Pull request by illicitonion**
The current random IDs are ugly and confusing when presented in UIs.
This makes them prettier and easier to read.

Also, disable non-automated registration of numeric IDs so that we don't
need to worry so much about people carving out our automated address
space and us needing to keep retrying ID registration.

### CODE REVIEW

File: synapse/handlers/register.py [hunk 1]
```diff
@@ -149,7 +162,7 @@ def register(
                     user_id = None
                     token = None
                     attempts += 1
-                    if attempts > 5:
+                    if attempts > 50:
```
Comment (richvdh) [inline synapse/handlers/register.py]: why is this change necessary?

Comment (illicitonion) [inline synapse/handlers/register.py]: We're now deterministically generating the things, so there's more likelihood of collisions.

On matrix.org we're ok with attempts > 2, but other servers may have more than 5 sequentially allocated numeric IDs

Comment (richvdh) [inline synapse/handlers/register.py]: But you're attempting to pick a starting point which is higher than any existing IDs? Or is it deliberate that you will re-issue any IDs which end up being a hole in the existing allocation? If the latter, 50 sounds like way too few.

Comment (illicitonion) [inline synapse/handlers/register.py]: It's filling holes.

I can bump to 5k or something if you'd prefer.

**Comment by richvdh** [comment]
Minor point: there are various comments which refer to "random" allocation of IDs. Please could these be fixed.

File: synapse/handlers/register.py (hunk 1 above)
Comment (richvdh) [inline synapse/handlers/register.py]: In that case, I'm not sure I'm entirely happy with the "try it and see if we get a SynapseError" approach to this. Is there really nothing else which could cause a SynapseError in that code block? I guess I'd prefer an approach which checked more directly for duplicates -- in which case I think you could remove the limit altogether.

File: synapse/storage/registration.py [hunk 2]
```diff
@@ -350,3 +352,27 @@ def _count_users(txn):
 
         ret = yield self.runInteraction("count_users", _count_users)
         defer.returnValue(ret)
+
+    @defer.inlineCallbacks
+    def find_next_generated_user_id(self):
```
Comment (richvdh) [inline synapse/storage/registration.py]: a doc-comment here could be useful to call out the fact that it will return the first hole in the sequence, rather than being relative to the highest allocation so far.

File: synapse/handlers/register.py (hunk 1 above)
Comment (illicitonion) [inline synapse/handlers/register.py]: Tidied up a bit. Reduced the scope of what can SynapseError. Removed the limit.

I could push the logic into a few different places (e.g. have the ID allocation be done in the store), but I don't care enough to make more invasive changes.

4a688a3

File: synapse/storage/registration.py (hunk 2 above)
Comment (illicitonion) [inline synapse/storage/registration.py]: Done fc3b7d8

**Comment by richvdh** [comment]
modulo the (unrelated?) sytest failure, lgtm
